tokio-sync/src/rwlock.rs

```diff
@@ -0,0 +1,213 @@
+//! An asynchronous `RwLock`-like type.
+//!
+//! This module provides [`RwLock`], a type that acts similarly to an asynchronous `RwLock`.
+//!
+//! This allows you to do something along the lines of:
+//!
+//! ```rust,no_run
+//! use tokio::sync::RwLock;
+//!
+//! #[tokio::main]
+//! async fn main() {
+//!     let data1 = RwLock::new(0);
+//!     let data2 = data1.clone();
+//!
+//!     tokio::spawn(async move {
+//!         let mut lock = data2.write().await;
+//!         *lock += 1;
+//!     });
+//!
+//!     let mut lock = data1.write().await;
+//!     *lock += 1;
+//! }
+//! ```
+//!
+//! [`RwLock`]: struct.RwLock.html
+//! [`RwLockReadGuard`]: struct.RwLockReadGuard.html
+//! [`RwLockWriteGuard`]: struct.RwLockWriteGuard.html
+
+use crate::task::AtomicWaker;
+use fnv::FnvHashMap;
+use futures_util::future::poll_fn;
+use std::ops;
+use std::sync::{
+    Arc, RwLock as StdRwLock, RwLockReadGuard as StdRwLockReadGuard,
+    RwLockWriteGuard as StdRwLockWriteGuard,
+};
+use std::task::Poll;
+
+/// An asynchronous readers writer lock usefull for protecting shared data
+///
+/// Each lock has a type parameter (`T`) which represents the data that it is protecting. The data
+/// can only be accessed through the RAII guards returned from `read` and `write`, which
+/// guarantees that the data is only ever modified when the rwlock is with write lock.
+#[derive(Debug)]
+pub struct RwLock<T> {
+    /// Pointer to the shared state
+    shared: Arc<Shared<T>>,
+
+    /// Pointer to the watcher's internal state
+    inner: Arc<AtomicWaker>,
```

Comment:
This looks suspicious to me. It looks like this is used in (at least) the read path, so multiple concurrent readers will get their wakers overwritten. Is that correct?

Reply:
yeah, it's what I highlighted in the solution description that I was not so sure about. Should  the list of watchers be instead composed of each read/write attempt, therefore each read/write guard having their own waker?
thanks!